### PR TITLE
Use ruff to lint Python code

### PR DESCRIPTION
**Pull request by cclauss**
[Ruff](https://docs.astral.sh/ruff) supports [over 800 lint rules](https://docs.astral.sh/ruff/rules/) and can be used to replace [Flake8](https://pypi.org/project/flake8/) (plus dozens of plugins), [isort](https://pypi.org/project/isort/), [pydocstyle](https://pypi.org/project/pydocstyle/), [yesqa](https://github.com/asottile/yesqa), [eradicate](https://pypi.org/project/eradicate/), [pyupgrade](https://pypi.org/project/pyupgrade/), and [autoflake](https://pypi.org/project/autoflake/), all while executing (in Rust) tens or hundreds of times faster than any individual tool.

`ruff check --output-format=github` provides intuitive GitHub Annotations to contributors...

image

Related to
* #488
* #489

@ulgens

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -62,3 +62,90 @@ strict_equality = true
...
+# Files not checked:
+#  - migrations: most of these are autogenerated and don't need a check
+#  - docs: contains autogenerated code that doesn't need a check
+exclude = [
+    "*/migrations/*",
+    "docs",
+]
+ignore = ["F401"]
+
+[tool.ruff.lint.mccabe]
```
Comment (dancergraham) [inline pyproject.toml:138]: question: why complexity 23 ?

File: pyproject.toml [hunk 2]
```diff
@@ -62,3 +62,90 @@ strict_equality = true
 [[tool.mypy.overrides]]
 module = ["django.*"]
 ignore_missing_imports = true
+
+[tool.ruff]
+line-length = 120
+target-version = "py38"
+
+[tool.ruff.lint]
```
Comment (dancergraham) [inline pyproject.toml:70]: question: why this set of lint rules?

File: waffle/templatetags/waffle_tags.py [hunk 3]
```diff
@@ -41,16 +41,15 @@ def render(self, context):
     def handle_token(cls, parser, token, kind, condition):
         bits = token.split_contents()
         if len(bits) < 2:
-            raise template.TemplateSyntaxError("%r tag requires an argument" %
-                                               bits[0])
+            raise template.TemplateSyntaxError(f"{bits[0]!r} tag requires an argument")
```
Comment (dancergraham) [inline waffle/templatetags/waffle_tags.py:44]: suggestion: use `repr(bits[0])` for readability in the spirit of https://peps.python.org/pep-0498/#s-r-and-a-are-redundant

**Review by dancergraham — COMMENTED** [review]
ruff is a great tool. This allows dev workflows with constant linting. It makes no significant difference to ci times as flake 8 runs in parallel with the tests and finishes first.

Django uses flake8 but I don't see a good reason why this library needs to wait for django to move

I've looked through the changes and can't see anything which should break.

**Comment by dancergraham** [comment]
I think the pre-commit.yaml file needs to be updated as well

File: pyproject.toml (hunk 1 above)
Comment (cclauss) [reply pyproject.toml:138]: Set it to 22 and see what happens.

File: pyproject.toml (hunk 2 above)
Comment (cclauss) [reply pyproject.toml:70]: These are the ones that currently pass.  Future PRs can uncomment rules that might be of interest.  We do not want this PR to contain so many changes that it slows the review process.

File: waffle/templatetags/waffle_tags.py (hunk 3 above)
Comment (cclauss) [reply waffle/templatetags/waffle_tags.py:44]: This PR is about automatic changes, not manual changes.  Turning on different rules may make other automatic changes.

File: pyproject.toml (hunk 1 above)
Comment (cclauss) [reply pyproject.toml:138]: Set it to 22 and see what happens...

% `pipx run ruff check --config mccabe.max-complexity=22`

File: pyproject.toml [hunk 4]
```diff
@@ -62,3 +62,90 @@ strict_equality = true
...
+    "docs",
+]
+ignore = ["F401"]
+
+[tool.ruff.lint.mccabe]
+max-complexity = 23
+
+[tool.ruff.lint.per-file-ignores]
+"docs/conf.py" = ["INP001"]
+"test_app/models.py" = ["DJ008"]  # FIXME
```
Comment (dancergraham) [inline pyproject.toml:143]: praise: nice to highlight this in the comment

File: pyproject.toml [hunk 5]
```diff
@@ -62,3 +62,90 @@ strict_equality = true
...
+"waffle/models.py" = ["DJ012", "PYI019"]  # FIXME
+
+[tool.ruff.lint.pylint]
+allow-magic-value-types = ["float", "int", "str"]
+max-args = 6  # default is 5
```
Comment (dancergraham) [inline pyproject.toml:148]: praise: nice to have the defaults mentioned here as a possible future target for refactoring

File: pyproject.toml [hunk 6]
```diff
@@ -62,3 +62,90 @@ strict_equality = true
...
+  "DTZ",    # flake8-datetimez
+  "E",      # pycodestyle errors
+  "F",      # Pyflakes
+  "FIX",    # flake8-fixme
+  "FLY",    # flynt
+  "G",      # flake8-logging-format
+  "ICN",    # flake8-import-conventions
+  "INP",    # flake8-no-pep420
+  "INT",    # flake8-gettext
+  "NPY",    # NumPy-specific rules
```
Comment (dancergraham) [inline pyproject.toml:86]: nit: this library doesn't use numpy or pandas and seems unlikely to start doing so so these could be omitted for a minor gain in readability

**Review by dancergraham — COMMENTED** [review]
I'm new around here so reluctant to personally approve such a big PR but it looks OK to me

**Review by clintonb — APPROVED** [review]

File: pyproject.toml (hunk 6 above)
Comment (cclauss) [reply pyproject.toml:86]: > this library doesn't use numpy or pandas and seems unlikely to start doing so

Famous last words.  You never know when code will be forked or when contributors will have new ideas.  Given that ruff can lint the entire CPython codebase in less than a third of a second with no cache, I take the stance that if a family of rules pass then run them.

**Pull request merged**